File: CSharp/SpaceVIL/SpaceVIL/Items/VerticalStack.cs
using SpaceVIL.Common;
using SpaceVIL.Core;
using System.Collections.Generic;

namespace SpaceVIL
{
    /// <summary>
    /// VerticalStack is a class that represents a line type container (vertical version). 
    /// VerticalStack groups items one after another using content alignment, margins, paddings, 
    /// sizes and size policies.
    /// VerticalStack implements SpaceVIL.Core.IHLayout.
    /// By default ability to get focus is disabled.
    /// <para/> VerticalStack cannot receive any events, 
    /// so VerticalStack is always in the ItemState.Base state.
    /// </summary>
    public class VerticalStack : Prototype, IVLayout
    {
        static int count = 0;

        private ItemAlignment _contentAlignment = ItemAlignment.Top;
        /// <summary>
        /// Setting content alignment within VerticalStack area. Default: ItemAlignment.Left.
        /// <para/> Supports only: ItemAlignment.Left, ItemAlignment.HCenter, ItemAlignment.Right.
        /// </summary>
        /// <param name="alignment">Content alignment as SpaceVIL.Core.ItemAlignment.</param>
        public void SetContentAlignment(ItemAlignment alignment)
        {
            if (alignment == ItemAlignment.Top || alignment == ItemAlignment.VCenter || alignment == ItemAlignment.Bottom)
                _contentAlignment = alignment;
        }
        /// <summary>
        /// Getting current content alignment.
        /// <para/> Can be: ItemAlignment.Left, ItemAlignment.HCenter, ItemAlignment.Right.
        /// </summary>
        /// <returns>Content alignment as SpaceVIL.Core.ItemAlignment.</returns>
        public ItemAlignment GetContentAlignment()
        {
            return _contentAlignment;
        }
        /// <summary>
        /// Default VerticalStack constructor.
        /// </summary>
        public VerticalStack()
        {
            SetItemName("VerticalStack_" + count);
            count++;
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.VerticalStack)));
            IsFocusable = false;

        }

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            return false;
        }
        /// <summary>
        /// Adding item to the VerticalStack. 
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public override void AddItem(IBaseItem item)
        {
            base.AddItem(item);
            UpdateLayout();
        }
        /// <summary>
        /// Inserting item to the VerticalStack container. 
        /// If the number of container elements is less than the index, 
        /// then the element is added to the end of the list.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        /// <param name="index">Index of insertion.</param>
        public override void InsertItem(IBaseItem item, int index)
        {
            base.InsertItem(item, index);
            UpdateLayout();
        }
        /// <summary>
        /// Removing the specified item from the VerticalStack container.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public override bool RemoveItem(IBaseItem item)
        {
            bool result = base.RemoveItem(item);
            if (result)
                UpdateLayout();
            return result;
        }
        /// <summary>
        /// Setting VerticalStack height. If the value is greater/less than the maximum/minimum 
        /// value of the height, then the height becomes equal to the maximum/minimum value.
        /// </summary>
        /// <param name="height">Weight of the VerticalStack.</param>
        public override void SetHeight(int height)
        {
            base.SetHeight(height);
            UpdateLayout();
        }
        /// <summary>
        /// Setting Y coordinate of the left-top corner of the VerticalStack.
        /// </summary>
        /// <param name="y">Y position of the left-top corner.</param>
        public override void SetY(int y)
        {
            base.SetY(y);
            UpdateLayout();
        }
        /// <summary>
        /// Updating all children positions (implementation of SpaceVIL.Core.IFreeLayout).
        /// </summary>
        public void UpdateLayout()
        {
            int total_space = GetHeight() - GetPadding().Top - GetPadding().Bottom;
            int free_space = total_space;
            int fixed_count = 0;
            int expanded_count = 0;

            List<int> maxHeightExpands = new List<int>();

            foreach (var child in GetItems())
            {
                if (child.IsVisible())
                {
                    if (child.GetHeightPolicy() == SizePolicy.Fixed)
                    {
                        fixed_count++;
                        free_space -= (child.GetHeight() + child.GetMargin().Top + child.GetMargin().Bottom);//
                    }
                    else
                    {
                        if (child.GetMaxHeight() < total_space)
                        {
                            maxHeightExpands.Add(child.GetMaxHeight() + child.GetMargin().Top + child.GetMargin().Bottom);
                        }
                        expanded_count++;
                    }
                }
            }
            free_space -= GetSpacing().Vertical * ((expanded_count + fixed_count) - 1);

            int height_for_expanded = 1;
            // if (expanded_count > 0 && free_space > expanded_count)
            //     height_for_expanded = free_space / expanded_count;

            maxHeightExpands.Sort();

            while (true)
            {
                if (expanded_count == 0)
                    break;

                if (free_space > expanded_count)
                    height_for_expanded = free_space / expanded_count;

                if (height_for_expanded <= 1 || maxHeightExpands.Count == 0)
                {
                    break;
                }

                if (height_for_expanded > maxHeightExpands[0])
                {
                    while (maxHeightExpands.Count > 0 && height_for_expanded > maxHeightExpands[0])
                    {
                        free_space -= maxHeightExpands[0];
                        maxHeightExpands.RemoveAt(0);
                        expanded_count--;
                    }
                }
                else
                {
                    break;
                }
            }

            int offset = 0;
            int startY = GetY() + GetPadding().Top;
            bool isFirstExpand = false;
            int diff = (free_space - height_for_expanded * expanded_count);
            if (expanded_count != 0 && diff > 0)
            {
                isFirstExpand = true;
            }

            if (expanded_count > 0 || _contentAlignment.Equals(ItemAlignment.Top))
            {
                foreach (var child in GetItems())
                {
                    if (child.IsVisible())
                    {
                        child.SetY(startY + offset + child.GetMargin().Top);//
                        if (child.GetHeightPolicy() == SizePolicy.Expand)
                        {
                            if (height_for_expanded - child.GetMargin().Top - child.GetMargin().Bottom < child.GetMaxHeight())//
                            {
                                child.SetHeight(height_for_expanded - child.GetMargin().Top - child.GetMargin().Bottom);//
                            }
                            else
                            {
                                // expanded_count--;
                                // free_space -= (child.GetHeight() + child.GetMargin().Top + child.GetMargin().Bottom);
                                // height_for_expanded = 1;
                                // if (expanded_count > 0 && free_space > expanded_count)
                                //     height_for_expanded = free_space / expanded_count;
                                child.SetHeight(child.GetMaxHeight()); //?
                            }

                            if (isFirstExpand)
                            {
                                if (child.GetHeight() + diff < child.GetMaxHeight())
                                {
                                    child.SetHeight(child.GetHeight() + diff);
                                    isFirstExpand = false;
                                }
                            }
                        }
                        offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//
                    }
                    //refactor
                    child.SetConfines();
                }
            }
            else
            {
                //for content alignment right
                if (_contentAlignment.Equals(ItemAlignment.Bottom))
                {
                    foreach (var child in GetItems())
                    {
                        if (child.IsVisible())
                        {
                            child.SetY(startY + offset + child.GetMargin().Top + free_space);
                            offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//

                            if (child.GetHeightPolicy() == SizePolicy.Expand)
                                child.SetHeight(child.GetMaxHeight());
                        }
                        child.SetConfines();
                    }
                }
                //for content alignment hcenter
                else if (_contentAlignment.Equals(ItemAlignment.VCenter))
                {
                    foreach (var child in GetItems())
                    {
                        if (child.IsVisible())
                        {
                            child.SetY(startY + offset + child.GetMargin().Top + free_space / 2);
                            offset += child.GetHeight() + GetSpacing().Vertical + child.GetMargin().Top + child.GetMargin().Bottom;//

                            if (child.GetHeightPolicy() == SizePolicy.Expand)
                                child.SetHeight(child.GetMaxHeight());
                        }
                        child.SetConfines();
                    }
                }
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/VerticalStack.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Core.IVLayout;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.SizePolicy;

/**
 * VerticalStack is a class that represents a line type container (vertical
 * version). VerticalStack groups items one after another using content
 * alignment, margins, paddings, sizes and size policies. VerticalStack
 * implements com.spvessel.spacevil.Core.IVLayout. By default ability to
 * get focus is disabled.
 * <p>
 * VerticalStack cannot receive any events, so VerticalStack is always in the
 * com.spvessel.spacevil.Flags.ItemStateType.BASE state.
 */
public class VerticalStack extends Prototype implements IVLayout {
    private static int count = 0;

    private ItemAlignment _contentAlignment = ItemAlignment.Top;

    /**
     * Setting content alignment within VerticalStack area.
     * <p>
     * Supports only: ItemAlignment.TOP, ItemAlignment.VCENTER,
     * ItemAlignment.BOTTOM.
     * <p>
     * Default: ItemAlignment.TOP.
     * 
     * @param alignment Content alignment as
     *                  com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setContentAlignment(ItemAlignment alignment) {
        if (alignment == ItemAlignment.Top || alignment == ItemAlignment.VCenter || alignment == ItemAlignment.Bottom)
            _contentAlignment = alignment;
    }

    /**
     * Getting current content alignment.
     * <p>
     * Can be: ItemAlignment.TOP, ItemAlignment.VCENTER, ItemAlignment.BOTTOM.
     * 
     * @return Content alignment as com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public ItemAlignment getContentAlignment() {
        return _contentAlignment;
    }

    /**
     * Default VerticalStack constructor.
     */
    public VerticalStack() {
        setItemName("verticalStack_" + count);
        count++;
        setStyle(DefaultsService.getDefaultStyle(VerticalStack.class));
        isFocusable = false;
    }

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        return false;
    }

    /**
     * Adding item to the VerticalStack.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     */
    @Override
    public void addItem(IBaseItem item) {
        super.addItem(item);
        updateLayout();
    }

    /**
     * Inserting item to the VerticalStack container. If the number of container
     * elements is less than the index, then the element is added to the end of the
     * list.
     * 
     * @param item  Item as com.spvessel.spacevil.Core.IBaseItem.
     * @param index Index of insertion.
     */
    @Override
    public void insertItem(IBaseItem item, int index) {
        super.insertItem(item, index);
        updateLayout();
    }

    /**
     * Removing the specified item from the VerticalStack container.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    @Override
    public boolean removeItem(IBaseItem item) {
        boolean result = super.removeItem(item);
        if (result)
            updateLayout();
        return result;
    }

    /**
     * Setting VerticalStack height. If the value is greater/less than the
     * maximum/minimum value of the height, then the height becomes equal to the
     * maximum/minimum value.
     * 
     * @param height Height of the VerticalStack.
     */
    @Override
    public void setHeight(int height) {
        super.setHeight(height);
        updateLayout();
    }

    /**
     * Setting Y coordinate of the left-top corner of the VerticalStack.
     * 
     * @param y Y position of the left-top corner.
     */
    @Override
    public void setY(int y) {
        super.setY(y);
        updateLayout();
    }

    /**
     * Updating all children positions (implementation of
     * com.spvessel.spacevil.Core.IVLayout).
     */
    public void updateLayout() {
        int total_space = getHeight() - getPadding().top - getPadding().bottom;
        int free_space = total_space;
        int fixed_count = 0;
        int expanded_count = 0;

        List<Integer> maxHeightExpands = new LinkedList<>();

        for (IBaseItem child : getItems()) {
            if (child.isVisible()) {
                if (child.getHeightPolicy() == SizePolicy.Fixed) {
                    fixed_count++;
                    free_space -= (child.getHeight() + child.getMargin().top + child.getMargin().bottom);//
                } else {
                    if (child.getMaxHeight() < total_space) {
                        maxHeightExpands.add(child.getMaxHeight() + child.getMargin().top + child.getMargin().bottom);
                    }
                    expanded_count++;
                }
            }
        }
        free_space -= getSpacing().vertical * ((expanded_count + fixed_count) - 1);

        int height_for_expanded = 1;
        // if (expanded_count > 0 && free_space > expanded_count)
        // height_for_expanded = free_space / expanded_count;

        Collections.sort(maxHeightExpands);

        while (true) {
            if (expanded_count == 0)
                break;

            if (free_space > expanded_count)
                height_for_expanded = free_space / expanded_count;

            if (height_for_expanded <= 1 || maxHeightExpands.size() == 0) {
                break;
            }

            if (height_for_expanded > maxHeightExpands.get(0)) {
                while (maxHeightExpands.size() > 0 && height_for_expanded > maxHeightExpands.get(0)) {
                    free_space -= maxHeightExpands.get(0);
                    maxHeightExpands.remove(0);
                    expanded_count--;
                }
            } else {
                break;
            }
        }

        int offset = 0;
        int startY = getY() + getPadding().top;
        boolean isFirstExpand = false;
        int diff = (free_space - height_for_expanded * expanded_count);
        if (expanded_count != 0 && diff > 0) {
            isFirstExpand = true;
        }

        if (expanded_count > 0 || _contentAlignment.equals(ItemAlignment.Top)) {
            for (IBaseItem child : getItems()) {
                if (child.isVisible()) {
                    child.setY(startY + offset + child.getMargin().top);//
                    if (child.getHeightPolicy() == SizePolicy.Expand) {
                        if (height_for_expanded - child.getMargin().top - child.getMargin().bottom < child
                                .getMaxHeight()) {
                            child.setHeight(height_for_expanded - child.getMargin().top - child.getMargin().bottom);
                        } else {
                            // expanded_count--;
                            // free_space -= (child.getHeight() + child.getMargin().top +
                            // child.getMargin().bottom);
                            // height_for_expanded = 1;
                            // if (expanded_count > 0 && free_space > expanded_count)
                            // height_for_expanded = free_space / expanded_count;
                            child.setHeight(child.getMaxHeight()); // ?
                        }

                        if (isFirstExpand) {
                            if (child.getHeight() + diff < child.getMaxHeight()) {
                                child.setHeight(child.getHeight() + diff);
                                isFirstExpand = false;
                            }
                        }
                    }
                    offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                            + child.getMargin().bottom;//
                }
                // refactor
                child.setConfines();
            }
        } else {
            if (_contentAlignment.equals(ItemAlignment.Bottom)) {
                for (IBaseItem child : getItems()) {
                    if (child.isVisible()) {
                        child.setY(startY + offset + child.getMargin().top + free_space);//
                        offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                                + child.getMargin().bottom;//

                        if (child.getHeightPolicy() == SizePolicy.Expand)
                            child.setHeight(child.getMaxHeight());
                    }
                    child.setConfines();
                }
            } else if (_contentAlignment.equals(ItemAlignment.VCenter)) {
                for (IBaseItem child : getItems()) {
                    if (child.isVisible()) {
                        child.setY(startY + offset + child.getMargin().top + free_space / 2);//
                        offset += child.getHeight() + getSpacing().vertical + child.getMargin().top
                                + child.getMargin().bottom;//

                        if (child.getHeightPolicy() == SizePolicy.Expand)
                            child.setHeight(child.getMaxHeight());
                    }
                    child.setConfines();
                }
            }
        }

    }
}